openest/generate/functions.py

```diff
@@ -443,6 +443,141 @@ def describe():
 
 
+class Product(calculation.Calculation):
+    """Product of two or more subcalculations
+
+    Parameters
+    ----------
+    subcalcs : Sequence of ``openest.generate.calculation.Calculation``
+    unshift : bool, optional
+    """
+
+    def __init__(self, subcalcs, unshift=True):
+        # Unit of result is product of input subcalc units.
+        units_product = [" * ".join([s.unitses[0] for s in subcalcs])]
+        if unshift:
+            fullunitses = [unit for calc in subcalcs for unit in calc.unitses]
+            super().__init__(units_product + fullunitses)
+        else:
+            super().__init__(units_product)
+
+        self.unshift = unshift
+        self.subcalcs = subcalcs
+
+
+    def format(self, lang, *args, **kwargs):
+        mains = []
+        elements = {}
+        alldeps = set()
+        for subcalc in self.subcalcs:
+            elements.update(subcalc.format(lang, *args, **kwargs))
+            mains.append(elements['main'])
+            alldeps.update(elements['main'].dependencies)
+
+        if lang in ['latex', 'julia']:
+            elements['main'] = FormatElement(' * '.join([main.repstr for main in mains]), list(alldeps))
+
+        formatting.add_label('product', elements)
+        return elements
+
+    def apply(self, region, *args, **kwargs):
+        """Apply calculation to all subcalculations
+
+        All parameters are passed to ``self.subcalc.apply()``.
+
+        Parameters
+        ----------
+        region : str
+        args
+        kwargs
+
+        Returns
+        -------
+        openest.generate.Calculation.ApplicationPassCall
+        """
+
+        def generate(year, results):
+            return np.prod([x[1] if x is not None else np.nan for x in results])
+
+        # Prepare the generator from our encapsulated operations
+        subapps = [subcalc.apply(region, *args, **kwargs) for subcalc in self.subcalcs]
+        return calculation.ApplicationPassCall(region, subapps, generate, unshift=self.unshift)
+
+    def column_info(self):
+        """Get column information of values output from this calculation.
+
+        Returns
+        -------
+        Sequence of dicts
+            Each dict contains:
+
+                ``"name"``
+                    Short-length title of this calculation
+
+                ``"title"``
+                    Long-length title of this calculation
+
+                ``"description"``
+                    Long-form description of this calculation
+        """
+        infoses = [subcalc.column_info() for subcalc in self.subcalcs]
+        title = 'Product of previous results'
+        description = 'Product of ' + ', '.join([infos[0]['title'] for infos in infoses])
+        fullinfos = [info for infos in infoses for info in infos]
+        return [dict(name='product', title=title, description=description)] + fullinfos
+
+    def enable_deltamethod(self):
+        raise NotImplementedError
```

Comment:
This should be a different exception. The problem is not that it has not been implemented, but that it is mathematically incorrect, and the user who has requested the delta method should be informed.

Reply:
Thank you for catching that. I wanted to raise a question on this but got sidetracked.

Resolved error type with 0dc4972.

I think this might reflect a larger issue in the framework. I have some ideas. I'll open a new issue on this.